pkg/kubelet/dockertools/manager.go

```diff
@@ -1844,64 +1898,21 @@ func (dm *DockerManager) SyncPod(pod *api.Pod, _ api.PodStatus, podStatus *kubec
 	if podStatus != nil {
 		podIP = podStatus.IP
 	}
+	podInfraContainerID := containerChanges.InfraContainerId
 
 	// If we should create infra container then we do it first.
-	podInfraContainerID := containerChanges.InfraContainerId
 	if containerChanges.StartInfraContainer && (len(containerChanges.ContainersToStart) > 0) {
-		glog.V(4).Infof("Creating pod infra container for %q", format.Pod(pod))
-		startContainerResult := kubecontainer.NewSyncResult(kubecontainer.StartContainer, PodInfraContainerName)
-		result.AddSyncResult(startContainerResult)
-		var msg string
-		podInfraContainerID, err, msg = dm.createPodInfraContainer(pod)
-		if err != nil {
-			startContainerResult.Fail(err, msg)
-			glog.Errorf("Failed to create pod infra container: %v; Skipping pod %q", err, format.Pod(pod))
-			return
-		}
-
 		setupNetworkResult := kubecontainer.NewSyncResult(kubecontainer.SetupNetwork, kubecontainer.GetPodFullName(pod))
 		result.AddSyncResult(setupNetworkResult)
-		if !usesHostNetwork(pod) {
-			// Call the networking plugin
-			err = dm.networkPlugin.SetUpPod(pod.Namespace, pod.Name, podInfraContainerID)
-			if err != nil {
-				// TODO: (random-liu) There shouldn't be "Skipping pod" in sync result message
-				message := fmt.Sprintf("Failed to setup network for pod %q using network plugins %q: %v; Skipping pod", format.Pod(pod), dm.networkPlugin.Name(), err)
-				setupNetworkResult.Fail(kubecontainer.ErrSetupNetwork, message)
-				glog.Error(message)
-
-				// Delete infra container
-				killContainerResult := kubecontainer.NewSyncResult(kubecontainer.KillContainer, PodInfraContainerName)
-				result.AddSyncResult(killContainerResult)
-				if delErr := dm.KillContainerInPod(kubecontainer.ContainerID{
-					ID:   string(podInfraContainerID),
-					Type: "docker"}, nil, pod, message); delErr != nil {
-					killContainerResult.Fail(kubecontainer.ErrKillContainer, delErr.Error())
-					glog.Warningf("Clear infra container failed for pod %q: %v", format.Pod(pod), delErr)
-				}
-				return
-			}
-
-			// Setup the host interface unless the pod is on the host's network (FIXME: move to networkPlugin when ready)
-			var podInfraContainer *docker.Container
-			podInfraContainer, err = dm.client.InspectContainer(string(podInfraContainerID))
-			if err != nil {
-				glog.Errorf("Failed to inspect pod infra container: %v; Skipping pod %q", err, format.Pod(pod))
-				result.Fail(err)
-				return
-			}
-
-			if dm.configureHairpinMode {
-				if err = hairpin.SetUpContainer(podInfraContainer.State.Pid, network.DefaultInterfaceName); err != nil {
-					glog.Warningf("Hairpin setup failed for pod %q: %v", format.Pod(pod), err)
-				}
-			}
-
-			// Overwrite the podIP passed in the pod status, since we just started the infra container.
-			podIP = dm.determineContainerIP(pod.Name, pod.Namespace, podInfraContainer)
+		newInfraContainerID, newPodIP, err := dm.setUpPodNetwork(pod)
```

Comment:
Does the removal of StartContainer for the pod infra container have API implications for things that watch kubelet events?  Seems like the event stream should get preserved, and I don't think it would be that hard.  Perhaps just emit the StartContainer event before calling setUpPodNetwork() and then have that function return a boolean indicating whether it got as far as starting networking or not, and use that to emit the SetupNetwork event?